### DISCUSSION TITLE
ASF Burst SLC products GMTSAR compatibility

### DISCUSSION THREAD

**Discussion by forrestfwilliams** [gmtsar/gmtsar#954] [General]
Hello All, my name is Forrest Williams and I work at the Alaska Satellite Facility, where I lead the Sentinel-1 burst access project. In this project, we're trying to provide our users with direct access to small subsets of consistently stacked SLC data (i.e., bursts) to make time-series and small area analyses simpler.

We have completed the burst archive and are now taking steps to make sure this data is compatible with all the major SAR processors, which is where GMTSAR comes in. I have created a Python package called [burst2safe](https://github.com/forrestfwilliams/burst2safe) that allows you to download a single burst or a set of bursts and reformat them as a SAFE file - the standard format for Sentinel-1 data. In the future, we hope to make this the default format for bursts.

I want to make sure the SAFE files produced by this tool are compatible with GMTSAR, and since I'm a GMTSAR novice, I could use some help with this.

Would anyone be willing to take burst2safe for a spin to see if the SAFE files produced by the tool work with GMTSAR? If not, I'd be happy to work together to make sure we make necessary changes!

**Comment by Xiaohua-Eric-Xu** [comment]
This is great! I think Eric Lindsey from University of New Mexico have already checked on the new product from ASF (https://github.com/gmtsar/gmtsar/issues/946). @ericlindsey This could be a solution.

**Reply by forrestfwilliams** [reply]
Thanks! Let me know if there is anything I can do to support your testing, or if any issues arise.

**Comment by AlexeyPechnikov** [comment]
@forrestfwilliams
> We have completed the burst archive 

Could you explain for which areas and what timeline? I recently checked, and I couldn’t find bursts for some areas at all, while for others, the data was available for not more than a couple of years.​

**Reply by forrestfwilliams** [reply]
@AlexeyPechnikov The burst archive should have the same temporal and spatial extents as the standard ASF Sentinel-1 collection. We just completed the archive last week, so if you haven't checked since then please check again. If there are areas/times where we are missing data, could you send me a Vertex search link so I can check it out?

**Comment by AlexeyPechnikov** [comment]
@forrestfwilliams 

Also, could you fix the annoying bug in your server bursts downloading API where the returned file size is invalid? It seems you don’t check the TIFF file size to validate the download. In my ASF bursts client, I validate the file size, and the size mismatch is a problem.

```
properties = result.geojson()['properties']
# properties['bytes'] is not an accurate file size but it looks about 40 kB smaller
if os.path.exists(tif_file) and os.path.getsize(tif_file) >= int(properties['bytes']):
...
```

**Reply by forrestfwilliams** [reply]
@AlexeyPechnikov the `bytes` property here actually represents the *compressed* size of the burst data. This is important information for extracting the burst data, but not useful for file checks. We can consider changing the name of this property if you think this is too confusing.

**Reply by AlexeyPechnikov** [reply]
@forrestfwilliams Sure, a name like ‘compressed_size’ would be much more informative, I suppose. Additionally, the actual size of the returned TIFF file is a very important attribute. Please add this too for the validation of the downloaded data. We don’t need to re-download the bursts when we can validate the size.

**Comment by ericlindsey** [comment]
This is exciting, thanks @forrestfwilliams! Unfortunately it did not work for me in my testing, but the error (segmentation fault) is not easy to debug. @Xiaohua-Eric-Xu I think this may be one for you to look into. 

I downloaded 2 consecutive bursts over an area in New Mexico, in case you want to test the same ones it is these:

S1_322391_IW1_20231017T010135_VV_D185-BURST
S1_322392_IW1_20231017T010138_VV_D185-BURST

These are from Path 151 ascending, subswath 1, both overlap POINT(-106.0522 34.4931) if you want to do a geographic search.

Then I installed burst2safe with conda, and ran it with the two burst names to create a simple 2-burst SAFE file - this worked successfully!

However, organize_files_tops_linux.csh failed with the following:

`make_s1a_tops /users/eol/test_burst/S1A_IW_SLC__1SSV_20231017T010135_20231017T010138_050798_061F3E_8244.SAFE/annotation/*iw1*vv*xml /users/eol/test_burst/S1A_IW_SLC__1SSV_20231017T010135_20231017T010138_050798_061F3E_8244.SAFE/measurement/*iw1*vv*tiff tmp2 0`
`Reading in 3378 lines of info from XML...`
`Writing 17 lines for orbit...`
`Segmentation fault`
` Couldn't open xml file:` 
` tmp2.PRM`
`couldn't open master.PRM`

It looks to me like it wrote the LED file correctly but failed when writing the PRM file. Possibly there is something missing in the XML file that it needs but the error message is not helpful. Xiaohua do you have any ideas? I've attached the xml file produced by burst2safe in case it is useful.

[s1a-iw1-slc-vv-20231017t010135-20231017t010141-050798-061f3e-001.xml.zip](https://github.com/user-attachments/files/16241328/s1a-iw1-slc-vv-20231017t010135-20231017t010141-050798-061f3e-001.xml.zip)

**Reply by ericlindsey** [reply]
@forrestfwilliams somewhat related, but when I download the burst images using my implementation of the ASF API (https://github.com/ericlindsey/sentinel_query_download) the images are received with non-unique names, like 1.tiff and 2.tiff so they all end up overwriting each other. And, I don't know how to get the XML file alongside, as it's not included in the .csv file returned by the query with dataset=SLC-BURST. 

Is there a different way I should be downloading them compared to the simple 'wget' I'm using? Or do I need to implement something to rename them after they are downloaded?

**Reply by forrestfwilliams** [reply]
@ericlindsey Yeah, the 1.tiff, 2.tiff naming scheme has annoyed me as well. With the transition to using SAFE as the default format, we can look at fixing this.

In general I'd recommend using [burst2safe](https://github.com/forrestfwilliams/burst2safe) to download burst data at this point, but if you want burst data in the original format I'd use the [asf_search](https://github.com/asfadmin/Discovery-asf_search) python module. This tool will correctly name the burst data on download (avoiding the 1.tiff issue) and also makes it easy to get the xml metadata as well.

Also, just FYI the xml data is accessible at the same url as the tiff data, just with a different filename extension (`.xml` instead of `.tiff`). For example, if a burst tiff url is:
```
https://sentinel1-burst.asf.alaska.edu/S1A_IW_SLC__1SDV_20240716T141630_20240716T141657_054787_06ABDE_95AE/IW1/VV/1.tiff
```
The associated xml file can be found at:
```
https://sentinel1-burst.asf.alaska.edu/S1A_IW_SLC__1SDV_20240716T141630_20240716T141657_054787_06ABDE_95AE/IW1/VV/1.xml
```

**Reply by AlexeyPechnikov** [reply]
@forrestfwilliams It would be ideal to completely resolve the Sentinel-1 SLC design issue and have single-burst scenes where the times in the scene name correspond to the burst capture time, and other identifiers match the original Sentinel-1 scene. This way, each SAFE file contains only one burst for a single polarization from a single subswath. In my implementation, the end time is just the burst start time plus 3 seconds. This approach allows downloading bursts without additional transformations and hosting them on fast static storage such as Amazon S3 or Google Cloud Buckets. Currently, interferogram calculation on a relatively slow Google Colab instance is sometimes faster than downloading the bursts. On a 32 or 64-core desktop computer, the complete SBAS and PSI processing is about 10 times faster than the data downloading! It’s absurd, isn’t it? Actually, all the problems are caused by the ESA Sentinel-1 SLC scene archive format.

**Reply by forrestfwilliams** [reply]
@AlexeyPechnikov This is what `burst2safe` is trying to accomplish. For example to get a single-burst SAFE for the granule above, you would use:

```
burst2safe S1_245714_IW1_20240716T141634_VV_95AE-BURST
```

We hope to put this tool behind an API in the future, but for right now you have to run it locally.

**Reply by AlexeyPechnikov** [reply]
@forrestfwilliams Unfortunately, `burst2safe` generates different scene names for every run and download the same data repeatedly, creating many duplicates with different names. This means there is no way to re-download missed bursts, and users need to manage possible duplicates with different names.

The PyGMTSAR ASF module allows you to download each burst just once, restart the download if needed, and download only new bursts when necessary. By the way, it seems you download the same scene bursts sequentially? That’s a bottleneck when we need multiple bursts from a single scene.

```
% ls -1

% time burst2safe S1_245714_IW1_20240716T141634_VV_95AE-BURST
Using granule search...
Found 1 burst(s).
Check burst group validity...
Downloading data...
Download complete.
Creating SAFE...
SAFE created!
burst2safe S1_245714_IW1_20240716T141634_VV_95AE-BURST  5.74s user 3.30s system 10% cpu 1:22.91 total
% ls -1
S1A_IW_SLC__1SSV_20240716T141634_20240716T141634_054787_06ABDE_87E0.SAFE

% time burst2safe S1_245714_IW1_20240716T141634_VV_95AE-BURST
Using granule search...
Found 1 burst(s).
Check burst group validity...
Downloading data...
Download complete.
Creating SAFE...
SAFE created!
burst2safe S1_245714_IW1_20240716T141634_VV_95AE-BURST  5.09s user 3.33s system 28% cpu 29.506 total
% ls -1
S1A_IW_SLC__1SSV_20240716T141634_20240716T141634_054787_06ABDE_87E0.SAFE
S1A_IW_SLC__1SSV_20240716T141634_20240716T141634_054787_06ABDE_AB1B.SAFE

% time burst2safe S1_245714_IW1_20240716T141634_VV_95AE-BURST
Using granule search...
Found 1 burst(s).
Check burst group validity...
Downloading data...
Download complete.
Creating SAFE...
SAFE created!
burst2safe S1_245714_IW1_20240716T141634_VV_95AE-BURST  6.39s user 3.36s system 5% cpu 3:11.41 total
% ls -1
S1A_IW_SLC__1SSV_20240716T141634_20240716T141634_054787_06ABDE_87E0.SAFE
S1A_IW_SLC__1SSV_20240716T141634_20240716T141634_054787_06ABDE_9581.SAFE
S1A_IW_SLC__1SSV_20240716T141634_20240716T141634_054787_06ABDE_AB1B.SAFE

...
```

**Comment by AlexeyPechnikov** [comment]
@ericlindsey

The downloaded TIFF files are okay, and we don’t need to modify them at all (unlike burst2safe). However, the XML file needs to be filtered and modified, including specifying the downloaded TIFF files’ offsets.

burst2safe bursts should work fine. I’ve compared the output, and it is almost the same as the PyGMTSAR ASF module downloaded bursts. The major difference is that the PyGMTSAR ASF module always creates single-burst scenes with persistent names, so we don’t need to re-download bursts repeatedly as burst2safe does for every run.

Your two bursts can be downloaded as follows:

```
python3.10 -m pip install pygmtsar
python3.10 -c "from pygmtsar import ASF; ASF('asf_user', 'asf_password').download_bursts('.', ['S1_322391_IW1_20231017T010135_VV_D185-BURSTT','S1_322392_IW1_20231017T010138_VV_D185-BURST'])"
```
<img width="1919" alt="image" src="https://github.com/user-attachments/assets/6b8a1d86-7a90-42e3-9f46-cf1e91f08c69">

You can find a complete example on Google Colab where the bursts are downloaded and processed using SBAS and PSI. The results are the same as the reference paper and using complete scenes processing. Since PyGMTSAR uses the GMTSAR Sentinel-1 preprocessor, GMTSAR processes the bursts well.

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1O3aZtZsTrQIldvCqlVRel13wJRLhmTJt?usp=sharing) PyGMTSAR SBAS and PSI Analyses: Lake Sarez Landslides, Tajikistan: https://colab.research.google.com/drive/1O3aZtZsTrQIldvCqlVRel13wJRLhmTJt?usp=sharing

<img src="https://github.com/AlexeyPechnikov/pygmtsar/assets/7342379/b507cad0-db7a-47e6-a679-f74631c5e840" width="100%">

<img width="1160" alt="image" src="https://github.com/user-attachments/assets/795af285-f222-436c-98dd-485f18634326">

<img width="1162" alt="image" src="https://github.com/user-attachments/assets/457db120-27ed-4bd3-9061-612251a5ec67">

<img width="1324" alt="image" src="https://github.com/user-attachments/assets/3d7ff0b4-beb9-44ad-aec9-a060f63547fd">

<img width="1916" alt="image" src="https://github.com/user-attachments/assets/c5c93a6a-01e3-47f5-95b5-b8ee9dd6dd5d">

**Comment by chintals** [comment]
This is great, I will test this out soon. 
@ericlindsey  If PRM creation is failing, it is possible the pop_burst in make_s1a_tops could be the culprit.

**Reply by chintals** [reply]
Yeah, I'm getting a segmentation error. Just an FYI, my test bursts are different from what Eric used. The LED file was created, but then I got a segmentation error.
![image](https://github.com/user-attachments/assets/a17ee17c-61ba-483a-a0c7-8d0605903cfc)

**Reply by Xiaohua-Eric-Xu** [reply]
You may try the fix here 5d1af99a3dda8511749bf7fe2b10b9c452537c08 in the master branch and see whether that can help get rid of the error.

**Comment by AlexeyPechnikov** [comment]
@chintals Try PyGMTSAR ASF module bursts downloading. In case you still experience any issues these are related to GMTSAR shell scripts because PyGMTSAR uses the same binaries for Sentinel-1 SLC preprocessing. As I’ve mentioned above, burst2safe producing almost the same bursts as PyGMTSAR ASF but you would compare them more detailed if the last works for you.
